---
title: "qbio490_fa23_final_CRC_R"
output: html_document
date: "2023-11-22"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

knitr::opts_knit$set(root.dir = normalizePath("/Users/abwilk/Documents/Fall 2023/QBIO 490/qbio_490_andrewwilk/analysis_data")) 
```

```{r}
# Downloading packages

if (!require("BiocManager", quietly = TRUE))
install.packages("BiocManager")
library(BiocManager)

if (!require("TCGAbiolinks", quietly = TRUE))
BiocManager::install("TCGAbiolinks")
library(TCGAbiolinks)

if (!require("maftools", quietly = TRUE))
BiocManager::install("maftools")
library(maftools)
```

```{r}
# Downloading clinical data

clin_query <- GDCquery(project = "TCGA-COAD",
                      data.category = "Clinical",
                      data.type = "Clinical Supplement",
                      data.format = 'BCR Biotab')
GDCdownload(clin_query)
clinical.BCRtab.all <- GDCprepare(clin_query)
clinic <- clinical.BCRtab.all$clinical_patient_coad[-c(1,2),]

colnames(clinic)[colnames(clinic) == "bcr_patient_barcode" ] <- "Tumor_Sample_Barcode"
```


```{r}
# Downloading mutation data
maf_query <- GDCquery(
                      project = "TCGA-COAD",
                      data.category = "Simple Nucleotide Variation",
                      access = "open",
                      data.type = "Masked Somatic Mutation",
                      workflow.type = "Aliquot Ensemble Somatic Variant Merging and Masking"
                      )

GDCdownload(maf_query)
maf <- GDCprepare(maf_query)
maf_object <- read.maf(maf = maf,
                        clinicalData = clinic,
                        isTCGA = TRUE)
```

```{r}
# Getting important maf dataframes

maf_clinical <- maf_object@clinical.data

maf_data <- maf_object@data
```

```{r}
# Downloading transcriptomic data

rna_query <- GDCquery(project ="TCGA-COAD",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")
GDCdownload(rna_query)
rna_se <- GDCprepare(rna_query)
```

```{r}
rna_clinical <- rna_se@colData
rna_clinical <- as.data.frame(rna_clinical)

treatments_mask <- ifelse(colnames(rna_clinical) == 'treatments', F,T)
rna_clinical <- rna_clinical[, treatments_mask]
primary_site_mask <- ifelse(colnames(rna_clinical) == 'primary_site', F,T)
rna_clinical <- rna_clinical[, primary_site_mask]
disease_type_mask <- ifelse(colnames(rna_clinical) == 'disease_type', F,T)
rna_clinical <- rna_clinical[, disease_type_mask]

rna_genes <- rna_se@rowRanges@elementMetadata
rna_genes <- as.data.frame(rna_genes)
rownames(rna_genes) <- rna_genes$gene_id


rna_counts <- rna_se@assays@data$unstranded
rna_counts <- as.data.frame(rna_counts)
rownames(rna_counts) <- rownames(rna_genes)
colnames(rna_counts) <- rownames(rna_clinical)
```

```{r}
dir.create("/Users/abwilk/Documents/Fall 2023/QBIO 490/qbio490_fa23_final_CRC/outputs")
```

```{r}
setwd("/Users/abwilk/Documents/Fall 2023/QBIO 490/qbio490_fa23_final_CRC/outputs")

current_dir <- getwd()

# Writing dataframes into csv's to be manipulated in python.

write.csv(clinic, file.path(current_dir, "clinical.csv"), row.names=FALSE)

write.csv(maf_clinical, file.path(current_dir, "maf_clinical.csv"), row.names=FALSE)

write.csv(maf_data, file.path(current_dir, "maf_data.csv"), row.names=FALSE)

write.csv(rna_clinical, file.path(current_dir, "rna_clinical.csv"), row.names=FALSE)

write.csv(rna_counts, file.path(current_dir, "rna_counts.csv"))

write.csv(rna_genes, file.path(current_dir, "rna_genes.csv"))
```

```{r}
### Before running this next section (DESeq2) must run Python script to generate csv files that will be loaded in as data frames ### 

rna_counts_path = '/Users/abwilk/Documents/Fall 2023/QBIO 490/qbio490_fa23_final_CRC/outputs/rna_counts2.csv'
rna_counts2 = read.csv(rna_counts_path)
rna_clinic_path = '/Users/abwilk/Documents/Fall 2023/QBIO 490/qbio490_fa23_final_CRC/outputs/rna_clinic2.csv'
rna_clinic2 = read.csv(rna_clinic_path)
```

```{r}
# make age_cat column. Remove NA's

rna_clinic2$age_category <- ifelse(rna_clinic2$age_at_diagnosis / 365 < 50, 'Early_Onset','Late_Onset')

# Factor relevant clinical variables

rna_clinic2$race <- factor(rna_clinic2$race)
rna_clinic2$age_category <- factor(rna_clinic2$age_category)
rna_clinic2$gender <- factor(rna_clinic2$gender)

# Get rid of low counts

rownames(rna_counts2) <- rna_counts2$Unnamed..0
rna_counts2$Unnamed..0 <- NULL

row_sums <- rowSums(rna_counts2)
low_counts_mask <- ifelse(row_sums < 10, F,T)

rna_counts2 <- rna_counts2[low_counts_mask,]
rna_genes2 <- rna_genes[low_counts_mask,]
```

```{r}
# Running DESeq2

if (!require("DESeq2", quietly = TRUE))
BiocManager::install("DESeq2")
library(DESeq2)

dds <- DESeqDataSetFromMatrix(countData = rna_counts2,
                             colData = rna_clinic2,
                             design = ~gender + race + age_category)
                             
dds_obj <- DESeq(dds)

resultsNames(dds_obj)

results <- results(dds_obj, format = "DataFrame", contrast = c("age_category", 'Early_Onset', 'Late_Onset'))

```

```{r}
# Volcano Plot

if (!require("EnhancedVolcano", quietly = TRUE))
BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)

EnhancedVolcano(results,
    lab = rna_genes2$gene_name,
    x = 'log2FoldChange',
    y = 'pvalue',
    pCutoff = 10e-3,
    FCcutoff = 1
    )
```
```{r}
# Finding most DE genes

rownames(listData) <- results@rownames

data <- data.frame(
  'Gene' = rna_genes2$gene_name,
  'log2FC' = results@listData$log2FoldChange,
  'p_adj' = results@listData$padj
)

masked_data <- na.omit(data)

sig_mask <- ifelse(abs(masked_data$log2FC) >= 1 & masked_data$p_adj <= 10e-3,T,F)

sig_genes <- masked_data[sig_mask,]$Gene

print(sig_genes)

# Moving this data into python
path = '/Users/abwilk/Documents/Fall 2023/QBIO 490/qbio490_fa23_final_CRC/outputs/DE_genes.csv'

write.csv(sig_genes, path, row.names=FALSE,quote=FALSE)
```

